### PR TITLE
AppDomainAssemblyCatalog correctly unloads the assemblies that were probed for any Nancy reference.

### PR DESCRIPTION
**Pull request by arturosevilla**
### Prerequisites

- [X] I have written a descriptive pull-request title
- [X] I have verified that there are no overlapping [pull-requests](https://github.com/NancyFx/Nancy/pulls) open
- [X] I have verified that I am following the Nancy [code style guidelines](https://github.com/NancyFx/Nancy/blob/45238076ad0b7f6ecabd6bae8469e30458d02efe/CONTRIBUTING.md#style-guidelines)
- [X] I have provided test coverage for my change (where applicable)

### Description

`AppDomainAssemblyCatalog`, in order to load all the assemblies that have a reference to Nancy, probes for the loaded assemblies as well as those that are in the "load path". The probing occurs by purposefully loading the assembly into a temporal `AppDomain` that is unloaded after the loading of useful assemblies is complete.

However this is accomplished by using `AppDomain.Load()`, and as specified by [MSDN](https://msdn.microsoft.com/en-us/library/36az8x58(v=vs.110).aspx), this is just a convenience method for `Assembly.Load()`. This means that the current code loads all the assemblies into both the inspection `AppDomain` and `AppDomain.Current`, when the code intent was clear that this is not the desired behavior.

Moreover, if the assembly is passed between application domains, because it is not serializable (`MarshalByRefObject`), it would be loaded as well into both domains, therefore the solution is to make a remote proxy probing instance which loads the assembly in the inspection domain and locally determines if the desired reference exists.

### CODE REVIEW

**Comment by arturosevilla** [comment]
I believe #2653 is breaking my PR.

**Comment by thecodejunkie** [comment]
You should rebase your PR and it should go away

**Comment by arturosevilla** [comment]
That did it 👍

File: src/Nancy/Helpers/ProxyNancyReferenceProber.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+{
+    using System;
+    using System.Reflection;
+    using Nancy.Extensions;
+
+    internal class ProxyNancyReferenceProber : MarshalByRefObject
+    {
+        public bool HasReference(AssemblyName assemblyNameForProbing, AssemblyName referenceAssemblyName)
+        {
+            var assemblyForInspection = Assembly.Load(assemblyNameForProbing);
```
Comment (khellang) [inline src/Nancy/Helpers/ProxyNancyReferenceProber.cs]: Can this be reflection only?

**Comment by khellang** [comment]
This is a really nice PR! :sparkles: Thanks :heart:

File: src/Nancy/Helpers/ProxyNancyReferenceProber.cs (hunk 1 above)
Comment (arturosevilla) [inline src/Nancy/Helpers/ProxyNancyReferenceProber.cs]: Ok, then yes. I tested it and it came out correctly, I just pushed these changes.

File: src/Nancy/Helpers/ProxyNancyReferenceProber.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#if !CORE
+namespace Nancy.Helpers
+{
+    using System;
+    using System.Reflection;
+    using Nancy.Extensions;
+
+    internal class ProxyNancyReferenceProber : MarshalByRefObject
```
Comment (arturosevilla) [inline src/Nancy/Helpers/ProxyNancyReferenceProber.cs]: @thecodejunkie I don't see which access modifier you're referring to? Besides making it `public`.

Comment (thecodejunkie) [inline src/Nancy/Helpers/ProxyNancyReferenceProber.cs]: I'd rather see it `private` than `internal`

Comment (arturosevilla) [inline src/Nancy/Helpers/ProxyNancyReferenceProber.cs]: @thecodejunkie But this is a class defined at the namespace level, you can't have a private class unless it's inside another class. Would you rather have this class put inside the `AppDomainAssemblyCatalog` class?

**Comment by thecodejunkie** [comment]
@khellang care to give this one last round of reviewing before we pull it in?

File: test/Nancy.Tests.MSBuild/Nancy.Tests.csproj [hunk 3]
```diff
@@ -651,6 +652,7 @@
     <Compile Include="..\Nancy.Tests\Unit\Responses\GenericFileResponseFixture.cs">
       <Link>Unit\Responses\GenericFileResponseFixture.cs</Link>
     </Compile>
+    <Compile Include="..\Nancy.Tests\Unit\AppDomainAssemblyCatalogFixture.cs" />
```
Comment (khellang) [inline test/Nancy.Tests.MSBuild/Nancy.Tests.csproj]: I think this should be referenced like the rest, using a `<Link>` element.

**Comment by thecodejunkie** [comment]
@khellang did you happen to kick the tires on it when you were working on it? Quick sanity check that everything is functioning as expected

**Comment by khellang** [comment]
@thecodejunkie Nope, but I'm assuming we have at least *some* test coverage there. Maybe that should be a billet point for 2.0? Do a coverage scan, see if we have big holes anywhere, and fix those?

**Comment by thecodejunkie** [comment]
@khellang it's mostly a runtime thing so hard to put good tests around it. Improving out coverage is always a good thing 👍

**Pull request merged**